fetchai/ledger/contract.py

```diff
@@ -22,11 +24,23 @@ def __init__(self, source: str):
         self._digest = _compute_digest(self._source)
         self._owner = None
 
-        # Quick and easy method to inspecting the contract source and generating a set of action and query names. To
-        # be replaced in the future with a more fault tolerant approach
-        ugly = ' '.join(map(lambda x: x.strip(), source.splitlines()))
-        self._actions = set(re.findall(r'@action function (\w+)\(', ugly))
-        self._queries = set(re.findall(r'@query function (\w+)\(', ugly))
+        # Etch parser for analysing contract
+        self._parser = EtchParser(self._source)
+
+        # Generate set of action and query entry points
+        entries = self._parser.entry_points(['init', 'action', 'query'])
+        if 'action' in entries:
```

Comment:
```suggestion
        self._actions = set(entries.get('action',[]))
```
Does this read a little better?

Reply:
Agreed and fixed